### PR TITLE
Refactor servo output to device, add per-servo output rate

### PR DESCRIPTION
**Pull request by CapnBry**
⚠️ The config struct is changed by this which will create a weird output mapping on the RX, be sure to set up the outputs via the webui after the first flashing ⚠️ 

## Remove latency from servo updates
The waveform generator built into the ESP8266 core only can update 1 output per global PWM edge, and will block if there's already any update queued. Our existing code attempted to work around this by not changing outputs more than once per 50Hz cycle, however that only lowers the first output's update latency. This meant output 2's update would block until the next edge of output 1 and so on. Updating all 5 servos could take anywhere from 5 to 100ms due to this architecture and probably where folks using head trackers could detect judder.

The new code can change all the outputs without blocking at all, and the new duty cycle (and refresh rate) is set on the next rising edge. Because `analogWrite` also uses this timer (there are only two timers), this means we can not analogWrite in the future without breaking the servo output. If that is needed the implementation can be updated to restore this functionality.

## Per-servo output rate
The existing code was locked at 50Hz update rate for the servos. This PR adds per-servo refresh rate configuration. 50Hz, 60Hz, 100Hz, 160Hz, 333Hz, and 400Hz is supported. In addition, there is also an "On/Off" mode that turns the output into a simple high / low switch, where anything up to 1500us input is LOW and anything above 1500us is HIGH.
image

## Narrow (750us) pulse width
It was requested to add 760us output mode as well, but just so nobody is happy I added it as 750us mode. This flag just halves the standard pulse width of 988-2012us -> 494-1006us.

## Refactor RX servo output to device
This removes the servo output code from rx_main and makes it a device. The purpose here was to force the initialization of the servo pins to OUTPUT / LOW on startup to reduce compatibility issues with ESCs that hate having the lines be left high (from pullups on some pins) until the TX connects. I do not believe this is going to have any effect though, because checking on the oscilloscope, the lines still go high for 70-120ms on startup before the device initialization occurs. I could not find any ESC that had a problem on any output pin with the old code. Yes I tried every ESC I own on all 5x output pins, repowering between each test. It was a fun weekend.

A servo manager class was also created to provide some abstraction for other future architecture implementations. This whole PR ends up being just 40 bytes more RAM and 30 bytes *less* code.

## Stop output PWM when entering webudate
Seemed like the right thing to do? The existing implementation just keeps choochin'.

## PWMPEX is now 6ch
The 7ch version didn't really seem to work, so the PWMPEX target is now 6ch, which does work we have a manufacturer interested in. The ADC scale for both the 6ch and 7ch targets is changed to 299 which is 33K:1K and supports up to 8S HV / 34.2V.

### CODE REVIEW

**Comment by JyeSmith** [comment]
> Because `analogWrite` also uses this timer (there are only two timers), this means we can not analogWrite in the future without breaking the servo output. If that is needed the implementation can be updated to restore this functionality.

Just making a note that the VTx dev uses analog write, but it will never be on a pwm esp8285 based rx due to the lack of pins.  The Pico (Frank) will have both but I imagine it does not have this limitation?

**Comment by CapnBry** [comment]
Yeah the Pico will use a different core so there's no issue there. We may need to rejigger the waveform_8266 header a little to get the VTx+RX to compile right, but that can be addressed on that PR.

**Review by pkendall64 — APPROVED** [review]
You are making work for me for the Unified target merge! But this is good stuff.
The servos seem a little more smooth now.

**Comment by CapnBry** [comment]
> You are making work for me for the Unified target merge! But this is good stuff. The servos seem a little more smooth now.

Oh I am sure that can go first and then I can just add this on top of it, that's a much bigger change and this is just moving a small bit of code, then some new files which should have no conflicts. Either way is fine with me, I just had to get this on the board before feature freeze. 

Wait until you see my FULL RES 12ch patch though ;)

**Comment by pkendall64** [comment]
> > You are making work for me for the Unified target merge! But this is good stuff. The servos seem a little more smooth now.
> 
> Oh I am sure that can go first and then I can just add this on top of it, that's a much bigger change and this is just moving a small bit of code, then some new files which should have no conflicts. Either way is fine with me, I just had to get this on the board before feature freeze.
> 
I think it's better to get all these other PRs merged and have the pain in one place 😄 

> Wait until you see my FULL RES 12ch patch though ;)

This will make soooo many people very happy!

**Comment by pkendall64** [comment]
@CapnBry I hope you don't mind, I merged master into this branch for you 😄

**Comment by CapnBry** [comment]
> @CapnBry I hope you don't mind, I merged master into this branch for you 😄

On the contrary, I appreciate this weekend gift, thanks!

File: src/lib/ServoOutput/waveform_8266.cpp [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+  uint32_t desiredHighCycles;  // Ideal waveform period to drive the error signal
+  uint32_t desiredLowCycles;   //
+  uint32_t nextHighLowUs;      // Waveform ideal (us) "on deck", waiting to be changed next cycle
+                               // packed into 32 bits to be atomic read/write, 65535us max
+  uint32_t lastEdge;           // Cycle when this generator last changed
+} Waveform;
+
+class WVFState {
+public:
+  Waveform waveform[17];        // State of all possible pins
```
Comment (schugabe) [inline src/lib/ServoOutput/waveform_8266.cpp:72]: Is 17 related to a hw limit or to the number of channels?

File: src/lib/ServoOutput/waveform_8266.cpp [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    timer1_isr_init();
+    timerRunning = false;
+  }
+}
+
+// Start up a waveform on a pin, or change the current one.  Will change to the new
+// waveform smoothly on next low->high transition.  For immediate change, stopWaveform()
+// first, then it will immediately begin.
+void startWaveform8266(uint8_t pin, uint32_t timeHighUS, uint32_t timeLowUS) {
+   if ((pin > 16) || isFlashInterfacePin(pin) || (timeHighUS == 0)) {
```
Comment (schugabe) [inline src/lib/ServoOutput/waveform_8266.cpp:131]: Same question: is this the max channel count?

File: src/html/scan.js [hunk 3]
```diff
@@ -13,15 +13,17 @@ function getPwmFormData()
...
         let failsafeField = _(`pwm_${ch}_fs`);
         let failsafe = failsafeField.value;
         if (failsafe > 2011) failsafe = 2011;
         if (failsafe < 988) failsafe = 988;
         failsafeField.value = failsafe;
 
-        let raw = (invert << 14) | (inChannel << 10) | (failsafe - 988);
-        //console.log(`PWM ${ch} input=${inChannel} fs=${failsafe} inv=${invert} raw=${raw}`);
+        let raw = (narrow << 19) | (mode << 15) | (invert << 14) | (inChannel << 10) | (failsafe - 988);
+        console.log(`PWM ${ch} mode=${mode} input=${inChannel} fs=${failsafe} inv=${invert} nar=${narrow} raw=${raw}`);
```
Comment (schugabe) [inline src/html/scan.js:26]: Do we need the debug output?

**Review by schugabe — APPROVED** [review]
Tested with different servos/frequencies and checked the waveform with a oscilloscope. I did not spot any issues :+1: 

There are some hard coded values for everything time based with the servo output (like 3000 for inverted or 988 for the failsafe offset) but I don't think that we will ever change the values so I don't think that a const would make sense in this case.

File: src/lib/ServoOutput/waveform_8266.cpp (hunk 1 above)
Comment (pkendall64) [reply src/lib/ServoOutput/waveform_8266.cpp:72]: GPIO0-16 on the esp8266 so I think theres a slot for each possible pin.

Comment (CapnBry) [reply src/lib/ServoOutput/waveform_8266.cpp:72]: That's correct. The code came from the esp8286 core except it only can change one servo per edge and has no way of accessing its internals to change that. I had to copy paste it over but then I cut out a lot of the other complexity. The RAM usage is about 10 more bytes overall due to each channel having to store its own next-queued-value. It really should be the max pin number we plan to use, but I figured that might be a hidden gotcha somewhere down the line if someone tried to use a pin higher than we use. Adding double indirection (a list of pins to waveform mapping) added a lot of code everywhere so I just left it how it was in the core at 17.

File: src/lib/ServoOutput/waveform_8266.cpp (hunk 2 above)
Comment (CapnBry) [reply src/lib/ServoOutput/waveform_8266.cpp:131]: Again code from the core had this so I left it.

File: src/html/scan.js (hunk 3 above)
Comment (CapnBry) [reply src/html/scan.js:26]: You're right. It was supposed to be commented back out, I'll fix.

**Comment by CapnBry** [comment]
> There are some hard coded values for everything time based with the servo output (like 3000 for inverted or 988 for the failsafe offset) but I don't think that we will ever change the values so I don't think that a const would make sense in this case.

The 3000 hard coded value is to flop the value around the centerpoint, which we know is 1500, We don't know where the end should be so it has to be mirrored around 1500, which is `(1500 - us) + 1500` which reduces to `3000 - us`. I'll add a comment so it is clearer.

I only wanted 10 bits to store the failsafe so it had to be 988-2011 for potential failsafe options. It used to fit in 16 bits per channel, but I ended up overflowing it by adding modes and narrow mode so I suppose that means it could be made 12-bit. I'm not sure our OTA will always support Extended Limits so I feel it is ok to limit it?

**Comment by CapnBry** [comment]
Do it, PK, do it!

**Pull request merged**